### ISSUE TITLE
Build distributed hyperparameter sweep engine with Grid/Random search and Celery chord orchestration

### ISSUE THREAD

**Issue by Sip4818** [Sip4818/TrainGrid#47]
## Overview

Implement a distributed hyperparameter sweep engine that creates N training runs with different hyperparameter combinations, dispatches them in parallel across Celery workers, and aggregates results to identify the best-performing configuration. Uses Celery `chord` / `group` for orchestration — no external dependencies (no Optuna).

---

## Files to Create/Modify

| File | Action | Purpose |
|------|--------|---------|
| `backend/shared/enums.py` | Edit | Add `SweepStatus` and `SearchStrategy` enums |
| `backend/shared/errors.py` | Edit | Add `SweepNotFoundError` |
| `backend/infrastructure/database/models.py` | Edit | Add `SweepModel` + add `sweep_id` and `project_id` FKs to `RunModel` |
| `backend/api/schemas/sweep.py` | Create | Pydantic schemas for sweep create/response |
| `backend/api/services/sweep_service.py` | Create | Business logic for creating and querying sweeps |
| `backend/workers/tasks/sweep_tasks.py` | Create | `aggregate_sweep_results` (chord callback) |
| `backend/workers/celery_app.py` | Edit | Register `sweep_tasks` in `imports` list |
| `backend/api/routers/sweeps.py` | Create | `POST /sweeps/`, `GET /sweeps/`, `GET /sweeps/{id}` |
| `backend/api/main.py` | Edit | Include `sweeps` router |
| `backend/api/core/exceptions.py` | Edit | Map `SweepNotFoundError` to HTTP 404 |
| `tests/workers/test_sweep_tasks.py` | Create | Tests for sweep aggregation logic |
| `tests/api/test_sweeps.py` | Create | Tests for sweep API endpoints |

---

## Step 1: `backend/shared/enums.py` — New enums

| Enum | Values | Purpose |
|------|--------|---------|
| `SweepStatus` | `PENDING`, `RUNNING`, `COMPLETED`, `FAILED` | Sweep lifecycle state |
| `SearchStrategy` | `GRID`, `RANDOM` | Search strategy type |

---

## Step 2: `backend/shared/errors.py` — New error

| Class | Extends | HTTP | Purpose |
|-------|---------|------|---------|
| `SweepNotFoundError` | `NotFoundError` | 404 | Sweep ID not found |

---

### `SweepModel`

| Column | Type | Purpose |
|--------|------|---------|
| `id` | `Integer`, PK | Primary key |
| `project_id` | `Integer`, FK → `projects.id` | Project scope (stored column) |
| `experiment_id` | `Integer`, FK → `experiments.id` | Experiment scope |
| `trainer_name` | `String` | Which trainer to use for all child runs |
| `search_space` | `JSON` | Hyperparameter search space definition |
| `strategy` | `SQLEnum(SearchStrategy)` | `"grid"` or `"random"` |
| `max_combinations` | `Integer`, nullable | For random search: max trials to sample. `None` for grid. |
| `status` | `SQLEnum(SweepStatus)` | Current sweep state |
| `best_run_id` | `Integer`, FK → `runs.id`, nullable | ID of the best run (set by aggregator) |
| `created_at` | `DateTime` | Timestamp |
| `started_at` | `DateTime`, nullable | When first child run started |
| `finished_at` | `DateTime`, nullable | When aggregator completed |

**Search space JSON format:**
```json
{
  "n_estimators": [100, 200, 300],
  "max_depth": [5, 10, 15],
  "learning_rate": [0.01, 0.1, 0.3]
}
```
Each key is a config field name, each value is a list of candidate values.

**Relationships:**
- `experiment` → `ExperimentModel`
- `runs` → list of `RunModel` (via `sweep_id` FK)
- `best_run` → `RunModel` (nullable)

### `RunModel` — Add columns

| Change | Detail |
|--------|--------|
| New column | `sweep_id: Column(Integer, ForeignKey("sweeps.id"), nullable=True)` — links run to its sweep |
| New relationship | `sweep = relationship("SweepModel", back_populates="runs")` |

Nullable — existing runs without a sweep remain valid.

---

### `SweepCreate` (request)

| Field | Type | Required | Purpose |
|-------|------|----------|---------|
| `experiment_id` | `int` | Yes | Experiment to scope the sweep |
| `project_id` | `int` | Yes | Project owning the experiment |
| `trainer_name` | `str` | Yes | Registered trainer name |
| `search_space` | `dict[str, list[Any]]` | Yes | Hyperparameter → list of candidate values |
| `strategy` | `str` | Yes | `"grid"` or `"random"` |
| `max_combinations` | `int \| None` | No | For random search: number of trials to sample |

**Validation:**
- `trainer_name` must be registered
- `search_space` values must be lists
- `max_combinations` required when `strategy == "random"`

### `Sweep` (response)

| Field | Type | Purpose |
|-------|------|---------|
| `id` | `int` | Sweep ID |
| `project_id` | `int` | Project |
| `experiment_id` | `int` | Experiment |
| `trainer_name` | `str` | Trainer |
| `search_space` | `dict` | Search space |
| `strategy` | `str` | Strategy |
| `max_combinations` | `int \| None` | Trial limit |
| `status` | `SweepStatus` | Current status |
| `best_run_id` | `int \| None` | Best run |
| `created_at` | `datetime` | Timestamp |
| `started_at` | `datetime \| None` | Start time |
| `finished_at` | `datetime \| None` | End time |
| `run_ids` | `list[int]` | IDs of all child runs |

### `SweepList` (list response)

| Field | Type | Purpose |
|-------|------|---------|
| `sweeps` | `list[Sweep]` | List of sweeps |

---

### `create_sweep(payload: SweepCreate) -> SweepModel`

1. Validate `experiment_id` + `project_id` (reuse scope validation pattern from `RunService`)
2. Validate `trainer_name` is registered via `trainer_registry.get()`
3. Validate `search_space` format (all values are non-empty lists)
4. Generate parameter combinations:
   - **Grid**: `itertools.product(*search_space.values())` → all combos
   - **Random**: sample `max_combinations` random combos from the product
5. Persist `SweepModel` with status `PENDING`
6. For each combination:
   - Build config: merge hyperparameter combo + `trainer_name` + `dataset_path`
   - Create `RunModel` with `status=PENDING`, `sweep_id=sweep.id`, `project_id=sweep.project_id`
   - Commit
7. Dispatch Celery chord: `chord(group(start_training_run.s(str(run_id)) for run_id in run_ids))(aggregate_sweep_results.s(sweep_id=str(sweep.id)))`
8. Update sweep status to `RUNNING`
9. Return sweep

### `get_sweep(sweep_id, project_id, experiment_id) -> SweepModel`

1. Fetch sweep, validate scope
2. Return with child run IDs

### `list_sweeps(project_id, experiment_id) -> list[SweepModel]`

1. Validate scope
2. Query all sweeps for the experiment

---

### `aggregate_sweep_results(run_ids: list[str], sweep_id: str)`

This is the chord callback that runs after all child training runs complete.

1. Open DB session
2. Fetch all child runs by IDs
3. Filter to `COMPLETED` runs (ignore `FAILED`)
4. Select best run: highest `metrics["accuracy"]` (single-objective for now)
5. Update `SweepModel.best_run_id` = best run's ID
6. Set sweep status to `COMPLETED` (or `FAILED` if no completed runs)
7. Set `finished_at`
8. Commit

**Note:** We reuse `start_training_run` directly for child runs — each is a standard training run. No wrapper needed.

### Chord dispatch pattern

```python
chord(
    group(start_training_run.s(str(run_id)) for run_id in run_ids)
)(aggregate_sweep_results.s(sweep_id=str(sweep.id)))
```

All `start_training_run` tasks run in parallel via `group`. `aggregate_sweep_results` runs only after ALL complete.

---

## Step 7: `backend/workers/celery_app.py` — Register sweep tasks

| Change | Detail |
|--------|--------|
| Add to `imports` | `"backend.workers.tasks.sweep_tasks"` |

---

## Step 8: `backend/api/routers/sweeps.py` — API endpoints

| Method | Path | Purpose |
|--------|------|---------|
| `POST /sweeps/` | Create and dispatch a sweep |
| `GET /sweeps/` | List all sweeps for an experiment |
| `GET /sweeps/{sweep_id}` | Get sweep details with child runs |

All scoped by `project_id` and `experiment_id` query params (same pattern as runs router).

---

## Step 9: `backend/api/main.py` — Include router

| Change | Detail |
|--------|--------|
| Add import | `from backend.api.routers import sweeps` |
| Add router | `app.include_router(sweeps.router)` |

---

## Step 10: `backend/api/core/exceptions.py` — Map error

| Change | Detail |
|--------|--------|
| Add handler | `SweepNotFoundError` → HTTP 404 with `{"code": "SWEEP_NOT_FOUND"}` |

---

### `tests/workers/test_sweep_tasks.py`

| Test | Verifies |
|------|----------|
| `test_aggregate_selects_best_run` | Given 3 completed runs with different accuracies, aggregator sets `best_run_id` to highest |
| `test_aggregate_handles_all_failed` | If all child runs failed, sweep status = `FAILED` |
| `test_aggregate_with_partial_failures` | Mix of completed/failed — best selected from completed only |
| `test_aggregate_sets_timestamps` | `finished_at` is set on the sweep |

### `tests/api/test_sweeps.py`

| Test | Verifies |
|------|----------|
| `test_create_grid_sweep` | Creates sweep, correct number of runs generated |
| `test_create_random_sweep` | Creates sweep with `max_combinations`, correct number of runs |
| `test_create_sweep_invalid_trainer` | Unknown trainer → `TrainerNotFoundError` → 422 |
| `test_create_sweep_invalid_search_space` | Non-list values → validation error |
| `test_list_sweeps` | Returns all sweeps for experiment |
| `test_get_sweep` | Returns sweep with child run IDs |
| `test_get_sweep_not_found` | Unknown sweep → 404 |

---

## How It Works (End-to-End)

**Example:** Grid search on `n_estimators: [100, 200]` and `max_depth: [5, 10]`

| Step | What happens |
|------|-------------|
| 1 | API receives `POST /sweeps/` with search space |
| 2 | Service generates 4 combinations: `[100,5], [100,10], [200,5], [200,10]` |
| 3 | Creates 1 `SweepModel` row + 4 `RunModel` rows (each with `status=PENDING`, `sweep_id` set) |
| 4 | Dispatches all 4 runs in parallel via Celery `group` |
| 5 | Each worker picks up one run, trains, evaluates, saves result |
| 6 | After all 4 finish, `aggregate_sweep_results` runs: picks best accuracy, sets `SweepModel.best_run_id` |

**Record counts:**

| Table | Rows created |
|-------|-------------|
| `sweeps` | 1 |
| `runs` | N (one per hyperparameter combination) |

---

## Design Decisions

| Decision | Rationale |
|----------|-----------|
| Reuse `start_training_run` | Each child run is a standard training run. No need to duplicate training logic. Sweep engine only handles orchestration. |
| `sweep_id` FK on `RunModel` | Nullable — backward-compatible. Links child runs to their sweep for querying and display. |
| `project_id` stored on `SweepModel` | Avoids ambiguity with derived property. Explicit FK for fast queries. |
| Chord over manual polling | Celery chord is idiomatic for "run callback after group completes". No polling, no timers. |
| Grid = cartesian product, Random = sample | Standard strategies. `itertools.product` for grid, `random.sample` for random. |
| Single-objective best selection | "Best" = highest accuracy for now. Pareto-optimal is a natural extension when multi-metric is added. |
| Search space as JSON dict | Simple, serializable, frontend-friendly. Keys are param names, values are candidate lists. |

---

## Execution Order

1. `backend/shared/enums.py`
2. `backend/shared/errors.py`
3. `backend/infrastructure/database/models.py`
4. `backend/api/schemas/sweep.py`
5. `backend/api/services/sweep_service.py`
6. `backend/workers/tasks/sweep_tasks.py`
7. `backend/workers/celery_app.py`
8. `backend/api/routers/sweeps.py`
9. `backend/api/main.py`
10. `backend/api/core/exceptions.py`
11. `tests/workers/test_sweep_tasks.py`
12. `tests/api/test_sweeps.py`
13. `./check.sh`
